File: engine/scheduler/cosmic_scheduler.py
#!/usr/bin/env python3
"""
Auto-Notion Cosmic Timing Engine
Aligns automation triggers with lunar cycles and auspicious transits
"""

import math
from datetime import datetime, timedelta
from typing import Dict, List, Optional
import logging

class CelestialEvent:
    """Represents a celestial event influencing timing"""
    def __init__(self, name: str, timestamp: float, duration_hours: float, sign: str):
        self.name = name
        self.timestamp = timestamp
        self.duration_hours = duration_hours
        self.sign = sign

class CosmicScheduler:
    """
    Institutional-Grade Cosmic Timing Engine
    Calculates lunar cycles and auspicious windows for optimal posting
    """
    
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        self.lunar_cycle_days = 29.53059
        # Epoch: 2024-01-11 11:57 UTC (New Moon)
        self.known_new_moon = datetime(2024, 1, 11, 11, 57)
        
    def get_moon_phase(self, date: datetime) -> float:
        """Calculate current moon phase (0.0 to 1.0)"""
        diff = date - self.known_new_moon
        phase = (diff.total_seconds() / (self.lunar_cycle_days * 24 * 3600)) % 1.0
        return phase
    
    def get_phase_name(self, phase: float) -> str:
        """Return human-readable phase name"""
        if phase < 0.05 or phase > 0.95: return "New Moon"
        if phase < 0.25: return "Waxing Crescent"
        if phase < 0.30: return "First Quarter"
        if phase < 0.45: return "Waxing Gibbous"
        if phase < 0.55: return "Full Moon"
        if phase < 0.75: return "Waning Gibbous"
        if phase < 0.80: return "Last Quarter"
        return "Waning Crescent"
    
    def is_auspicious(self, date: datetime) -> bool:
        """
        Deterministic check for auspicious windows
        Based on lunar phase and solar transit
        """
        phase = self.get_moon_phase(date)
        
        # Rule 1: Full Moon and Waxing phases are generally auspicious for growth
        # Rule 2: Avoid New Moon (low energy) and Eclipse windows (chaotic)
        
        if 0.05 < phase < 0.95: # Not a New Moon
            if 0.45 < phase < 0.55: # Full Moon Peak
                return True
            if 0.10 < phase < 0.40: # Growth phase
                return True
        
        # Rule 3: High-noon correlation (Brahma Muhurta - simplified)
        hour = date.hour
        if 4 <= hour <= 6: return True # Dawn
        if 18 <= hour <= 20: return True # Sunset
        
        return False
    
    def get_next_auspicious_window(self, start_date: datetime) -> datetime:
        """Find the next deterministic auspicious window"""
        current = start_date
        # Check every 30 minutes for the next 48 hours
        for _ in range(96):
            if self.is_auspicious(current):
                return current
            current += timedelta(minutes=30)
        return start_date # Fallback
    
    def align_schedule(self, content_plan: List[Dict]) -> List[Dict]:
        """Align a content plan with cosmic windows"""
        self.logger.info("Aligning schedule with Cosmic Timing Engine...")
        
        aligned_plan = []
        last_date = datetime.now()
        
        for item in content_plan:
            auspicious_time = self.get_next_auspicious_window(last_date + timedelta(hours=4))
            
            item['scheduled_date'] = auspicious_time.strftime("%Y-%m-%d")
            item['scheduled_time'] = auspicious_time.strftime("%H:%M")
            item['cosmic_metadata'] = {
                'moon_phase': round(self.get_moon_phase(auspicious_time), 4),
                'phase_name': self.get_phase_name(self.get_moon_phase(auspicious_time)),
                'is_peak': 0.45 < self.get_moon_phase(auspicious_time) < 0.55
            }
            
            aligned_plan.append(item)
            last_date = auspicious_time
            
        return aligned_plan

if __name__ == "__main__":
    scheduler = CosmicScheduler()
    now = datetime.now()
    phase = scheduler.get_moon_phase(now)
    print(f"Current Date: {now}")
    print(f"Moon Phase: {phase:.4f} ({scheduler.get_phase_name(phase)})")
    print(f"Is Auspicious: {scheduler.is_auspicious(now)}")
    
    next_up = scheduler.get_next_auspicious_window(now)
    print(f"Next Cosmic Window: {next_up}")

File: apps/dashboard/src/utils/cosmicScheduler.ts
// Cosmic Scheduler Utility
// Calculates lunar phases and auspicious timing windows

export interface LunarPhase {
    name: string;
    emoji: string;
    percentage: number;
    isAuspicious: boolean;
}

export interface CosmicWindow {
    start: Date;
    end: Date;
    type: 'lunar' | 'eclipse' | 'transit';
    description: string;
    auspiciousness: 'high' | 'medium' | 'low';
}

/**
 * Calculate current lunar phase
 * Uses simplified algorithm based on synodic month (29.53 days)
 */
export function getCurrentLunarPhase(): LunarPhase {
    const knownNewMoon = new Date('2000-01-06T18:14:00Z'); // Reference new moon
    const synodicMonth = 29.530588853; // Days
    const now = new Date();

    const daysSinceNew = (now.getTime() - knownNewMoon.getTime()) / (1000 * 60 * 60 * 24);
    const currentPhase = (daysSinceNew % synodicMonth) / synodicMonth;

    let phaseName: string;
    let emoji: string;
    let isAuspicious: boolean;

    if (currentPhase < 0.125) {
        phaseName = 'New Moon';
        emoji = '🌑';
        isAuspicious = true;
    } else if (currentPhase < 0.25) {
        phaseName = 'Waxing Crescent';
        emoji = '🌒';
        isAuspicious = true;
    } else if (currentPhase < 0.375) {
        phaseName = 'First Quarter';
        emoji = '🌓';
        isAuspicious = false;
    } else if (currentPhase < 0.5) {
        phaseName = 'Waxing Gibbous';
        emoji = '🌔';
        isAuspicious = true;
    } else if (currentPhase < 0.625) {
        phaseName = 'Full Moon';
        emoji = '🌕';
        isAuspicious = true;
    } else if (currentPhase < 0.75) {
        phaseName = 'Waning Gibbous';
        emoji = '🌖';
        isAuspicious = false;
    } else if (currentPhase < 0.875) {
        phaseName = 'Last Quarter';
        emoji = '🌗';
        isAuspicious = false;
    } else {
        phaseName = 'Waning Crescent';
        emoji = '🌘';
        isAuspicious = false;
    }

    return {
        name: phaseName,
        emoji,
        percentage: currentPhase,
        isAuspicious
    };
}

/**
 * Get next auspicious cosmic windows for content posting
 */
export function getAuspiciousWindows(days: number = 7): CosmicWindow[] {
    const windows: CosmicWindow[] = [];
    const now = new Date();

    // Generate windows based on lunar phases
    for (let i = 0; i < days; i++) {
        const date = new Date(now);
        date.setDate(date.getDate() + i);

        const hour = date.getHours();

        // Morning window (5-7 AM) - High frequency posting
        if (hour >= 5 && hour < 7) {
            windows.push({
                start: new Date(date.setHours(5, 0, 0, 0)),
                end: new Date(date.setHours(7, 0, 0, 0)),
                type: 'lunar',
                description: 'Morning Awakening - High Consciousness Window',
                auspiciousness: 'high'
            });
        }

        // Evening window (6-9 PM) - Contemplation time
        if (hour >= 18 && hour < 21) {
            windows.push({
                start: new Date(date.setHours(18, 0, 0, 0)),
                end: new Date(date.setHours(21, 0, 0, 0)),
                type: 'lunar',
                description: 'Evening Reflection - Integration Window',
                auspiciousness: 'high'
            });
        }
    }

    // Add special lunar windows
    const phase = getCurrentLunarPhase();
    if (phase.isAuspicious) {
        windows.push({
            start: now,
            end: new Date(now.getTime() + 24 * 60 * 60 * 1000),
            type: 'lunar',
            description: `${phase.name} - Amplified Manifestation Period`,
            auspiciousness: 'high'
        });
    }

    return windows.sort((a, b) => a.start.getTime() - b.start.getTime());
}

/**
 * Check if current time is within an auspicious window
 */
export function isAuspiciousNow(): boolean {
    const phase = getCurrentLunarPhase();
    const hour = new Date().getHours();

    // Auspicious lunar phase
    if (phase.isAuspicious) return true;

    // Auspicious hours (sunrise/sunset approximation)
    if ((hour >= 5 && hour < 7) || (hour >= 18 && hour < 21)) return true;

    return false;
}

/**
 * Get cosmic timing recommendation for posting
 */
export function getTimingRecommendation(): {
    shouldPost: boolean;
    reason: string;
    nextWindow?: Date;
} {
    const isAuspicious = isAuspiciousNow();
    const windows = getAuspiciousWindows(1);
    const nextWindow = windows.find(w => w.start > new Date());

    if (isAuspicious) {
        return {
            shouldPost: true,
            reason: 'Current time aligns with cosmic flow - ideal for high-frequency messaging'
        };
    }

    return {
        shouldPost: false,
        reason: 'Low-frequency period - content may not resonate at full potential',
        nextWindow: nextWindow?.start
    };
}
